src/EventStore.Core/TransactionLog/Chunks/TFChunk/WriterWorkItem.cs

```diff
@@ -1,90 +1,74 @@
-		}
+	internal sealed class WriterWorkItem : Disposable {
+		public const int BufferSize = 8192;
 
-		public long StreamPosition {
-			get { return _workingStream.Position; }
-		}
+		public Stream WorkingStream { get; private set; }
 
 		private readonly Stream _fileStream;
 		private UnmanagedMemoryStream _memStream;
-		private Stream _workingStream;
 
 		public readonly MemoryStream Buffer;
 		public readonly BinaryWriter BufferWriter;
 		public readonly HashAlgorithm MD5;
 
-		public WriterWorkItem(Stream fileStream, UnmanagedMemoryStream memStream, HashAlgorithm md5) {
-			_fileStream = fileStream;
-			_memStream = memStream;
-			_workingStream = (Stream)fileStream ?? memStream;
-			Buffer = new MemoryStream(8192);
-			BufferWriter = new BinaryWriter(Buffer);
+		public unsafe WriterWorkItem(nint memoryPtr, int length, HashAlgorithm md5) {
+			WorkingStream = _memStream = new UnmanagedMemoryStream((byte*)memoryPtr, length, length, FileAccess.ReadWrite);
+			Buffer = new(BufferSize);
+			BufferWriter = new(Buffer);
+			MD5 = md5;
+		}
+
+		public WriterWorkItem(SafeFileHandle handle, HashAlgorithm md5, bool unbuffered) {
+			WorkingStream = _fileStream = unbuffered
+				? handle.AsUnbufferedStream(FileAccess.ReadWrite)
+				: new BufferedStream(handle.AsUnbufferedStream(FileAccess.ReadWrite), BufferSize);
+			Buffer = new(BufferSize);
+			BufferWriter = new(Buffer);
 			MD5 = md5;
 		}
 
 		public void SetMemStream(UnmanagedMemoryStream memStream) {
 			_memStream = memStream;
-			if (_fileStream == null)
-				_workingStream = memStream;
+			if (_fileStream is null)
+				WorkingStream = memStream;
 		}
 
 		public void AppendData(byte[] buf, int offset, int len) {
 			// as we are always append-only, stream's position should be right here
-			if (_fileStream != null)
-				_fileStream.Write(buf, 0, len);
+			_fileStream?.Write(buf, offset, len);
+
 			//MEMORY
-			var memStream = _memStream;
-			if (memStream != null)
-				memStream.Write(buf, 0, len);
+			_memStream?.Write(buf, offset, len);
 		}
 
 		public void ResizeStream(int fileSize) {
-			if (_fileStream != null)
-				_fileStream.SetLength(fileSize);
-			var memStream = _memStream;
-			if (memStream != null)
-				memStream.SetLength(fileSize);
+			_fileStream?.SetLength(fileSize);
+			_memStream?.SetLength(fileSize);
 		}
 
-		public void Dispose() {
-			if (_fileStream != null)
-				_fileStream.Dispose();
+		protected override void Dispose(bool disposing) {
+			if (disposing) {
+				_fileStream?.Dispose();
+				DisposeMemStream();
+				Buffer.Dispose();
+				BufferWriter.Dispose();
+				MD5.Dispose();
+			}
 
-			DisposeMemStream();
+			base.Dispose(disposing);
 		}
 
-		public void FlushToDisk() {
-			if (_fileStream == null) return;
-			var fs = _fileStream as FileStream;
-			if (fs != null)
-				fs.FlushToDisk(); //because of broken flush in filestream in 3.5
-			else {
-				_fileStream.Flush();
-			}
-		}
+		public void FlushToDisk() => _fileStream?.Flush();
```

Comment:
the old code used to set `flushToDisk: true` in the eventual filestream flush and call, and on windows also call FlushFileBuffers

does what we have now do the same as the old filestream 'flushToDisk: true` behaviour

is FlushFileBuffers definitely not necessary (we still have a safeFileHandle if we want to call it)

Reply:
`Stream` wrapper over file handle does exactly what we expected: https://github.com/dotnet/dotNext/blob/master/src/DotNext.IO/IO/UnbufferedFileStream.cs#L48